main/src/ca/uwaterloo/flix/api/lsp/provider/InlayHintProvider.scala

```diff
@@ -13,15 +13,86 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package ca.uwaterloo.flix.api.lsp.provider
 
-import ca.uwaterloo.flix.api.lsp.{InlayHint, Range}
+package ca.uwaterloo.flix.api.lsp.provider
+import ca.uwaterloo.flix.api.lsp.acceptors.{FileAcceptor}
+import ca.uwaterloo.flix.language.ast.shared.SymUse
+import ca.uwaterloo.flix.language.ast.TypedAst.{Expr, Root}
+import ca.uwaterloo.flix.language.ast.Type
+import ca.uwaterloo.flix.language.ast.SourceLocation
+import ca.uwaterloo.flix.api.lsp.{Consumer, InlayHint, InlayHintKind, Position, Range}
+import ca.uwaterloo.flix.api.lsp.Visitor
+import scala.collection.immutable.Map
 
 object InlayHintProvider {
+  private val EnableEffectHints: Boolean = false
 
-  def getInlayHints(uri: String, range: Range): List[InlayHint] = {
-    // We currently do not support any inlay hints.
-    Nil
+  def getInlayHints(uri: String, range: Range)(implicit root: Root): List[InlayHint] = {
+    var inlayHints: List[InlayHint] = Nil
+    if(EnableEffectHints) {
+      var effects: Set[(Type, SourceLocation)] = (getOpSymUses(uri) ++ getDefSymUses(uri)).filter {
+        case (eff, loc) => eff match {
+          case Type.Pure => false
+          case _ => true
+        }
+      }
+      var maxColWidth: Int = Math.min(120, effects.map { case (_, loc) => loc.endCol }.maxOption.getOrElse(120))
+      var lineToEffectsMap: Map[Int, Set[Type]] = Map.empty
+      effects.foreach { case (eff, loc) =>
+        lineToEffectsMap = lineToEffectsMap.updated(loc.endLine, lineToEffectsMap.getOrElse(loc.endLine, Set.empty) + eff)
+      }
+      inlayHints = mkHintsFromEffects(lineToEffectsMap, maxColWidth)
+    }
+    inlayHints
   }
 
+  def getOpSymUses(uri: String)(implicit root: Root): Set[(Type, SourceLocation)] = {
+    var opSymUses: Set[(Type, SourceLocation)] = Set.empty
+    object opSymUseConsumer extends Consumer {
+      override def consumeExpr(expr: Expr): Unit = {
+        expr match {
+          case Expr.Do(_, _, _, eff, loc) =>
+            opSymUses += ((eff, loc))
+          case _ => ()
+        }
+      }
+    }
+    Visitor.visitRoot(root, opSymUseConsumer, FileAcceptor(uri))
+    opSymUses
+  }
+
+  def getDefSymUses(uri: String)(implicit root: Root): Set[(Type, SourceLocation)] = {
+    var defSymUses: Set[(Type, SourceLocation)] = Set.empty
+    object defSymUseConsumer extends Consumer {
+      override def consumeExpr(expr: Expr): Unit = {
+        expr match {
+          case Expr.ApplyDef(_, _, _, _, eff, loc) =>
+            defSymUses += ((eff, loc))
+          case _ => ()
+        }
+      }
+    }
+    Visitor.visitRoot(root, defSymUseConsumer, FileAcceptor(uri))
+    defSymUses
+  }
+
+  def mkHintsFromEffects(lineToEffectsMap: Map[Int, Set[Type]], maxColWidth: Int): List[InlayHint] = {
```

Comment:
Documentation on what `maxColWidth` precisely means. (E.g. does it include the hint?)